File: server/evals/lib/fixtures.js
// 사례 로더. 조정용 풀과 보류 풀을 분리해 읽고, 보류 사례가 조정용 인물을 쓰면 거부한다.
import { readFileSync, readdirSync } from 'node:fs';
import { dirname, join } from 'node:path';
import { fileURLToPath } from 'node:url';

const ROOT = join(dirname(fileURLToPath(import.meta.url)), '..');
const read = (p) => JSON.parse(readFileSync(join(ROOT, p), 'utf8'));
const strip = (pool) => Object.fromEntries(Object.entries(pool).filter(([k]) => !k.startsWith('_')));

export const TUNING_POOL = strip(read('fixtures/participants.json'));
export const HELDOUT_POOL = strip(read('fixtures/participants_heldout.json'));
// 프롬프트 동결 후 새로 만든 보류 풀. 기존 보류(H01~H07)는 이미 결과를 본 뒤라 더 이상
// 신선하지 않으므로 파일을 따로 둔다.
export const FRESH_POOL = strip(read('fixtures/participants_fresh.json'));
const POOLS = { tuning: TUNING_POOL, heldout: HELDOUT_POOL, fresh: FRESH_POOL };

/**
 * 후보를 고정된 순서로 섞는다.
 *
 * 사례 파일에 적은 후보 순서는 사람이 읽기 좋게 '기대하는 상위 후보부터' 적혀 있다. 그런데 규칙
 * 기준선은 동점을 입력 순서로 깨기 때문에, 그대로 두면 기준선이 아무 판단 없이 1순위를 맞히는
 * 착시가 생긴다. 시드를 고정한 해시 정렬로 순서에서 정답 신호를 지우되, 실행할 때마다 같은
 * 순서가 나오게 한다.
 *
 * 시드는 case_id 가 아니라 shuffle_group 이다. 교류 의도만 바꾼 비교 쌍(E01/E02/E07,
 * H01/H02)은 후보 제시 순서까지 같아야 한다. 순서가 다르면 순위가 바뀐 원인이 의도 변경인지
 * 배치 맥락과 동점 처리의 차이인지 갈라낼 수 없다.
 */
function fnv1a(text) {
  let hash = 0x811c9dc5;
  for (let i = 0; i < text.length; i += 1) {
    hash ^= text.charCodeAt(i);
    hash = Math.imul(hash, 0x01000193) >>> 0;
  }
  return hash;
}

function shuffleDeterministically(candidates, seed) {
  return [...candidates].sort((a, b) => {
    const ha = fnv1a(`${seed}:${a.user_id}`);
    const hb = fnv1a(`${seed}:${b.user_id}`);
    return ha - hb || a.user_id.localeCompare(b.user_id);
  });
}

function resolve(entry, pool, caseId, role) {
  const base = pool[entry.ref];
  if (!base) throw new Error(`${caseId}: ${role} 참조 '${entry.ref}'가 해당 풀에 없습니다.`);
  const { ref, ...overrides } = entry;
  return { ...base, ...overrides };
}

/** 사례 파일을 실행 가능한 요청 + 기대값으로 펼친다. */
export function loadCase(caseId) {
  const body = read(`fixtures/cases/${caseId}.json`);
  // 사례는 선언한 풀만 쓴다. 섞이면 보류 성격이 사라진다.
  const pool = POOLS[body.pool] ?? TUNING_POOL;
  const viewer = resolve(body.request.viewer, pool, caseId, 'viewer');
  const authored = body.request.candidates.map((c) => resolve(c, pool, caseId, 'candidate'));
  const candidates = shuffleDeterministically(authored, body.shuffle_group ?? caseId);
  return {
    ...body,
    request: {
      viewer,
      candidates,
      policy: body.request.policy ?? {},
      fault: body.request.fault ?? null,
      case_id: caseId,
      authored_candidate_order: authored.map((c) => c.user_id),
      shuffle_group: body.shuffle_group ?? caseId,
    },
  };
}

export function listCases(set) {
  return readdirSync(join(ROOT, 'fixtures/cases'))
    .filter((f) => f.endsWith('.json'))
    .map((f) => f.replace(/\.json$/, ''))
    .filter((id) => !set || read(`fixtures/cases/${id}.json`).set === set)
    .sort();
}

/** 조회자·후보가 실제로 쓴 원문 전체. 근거 검사의 유일한 출처다. */
export function sourceTexts(request) {
  const people = [request.viewer, ...request.candidates];
  return {
    all: people.map((p) => `${p.self_description} ${p.connection_intent}`).join(' '),
    byId: Object.fromEntries(
      request.candidates.map((c) => [
        c.user_id,
        {
          candidate: `${c.self_description} ${c.connection_intent}`,
          viewer: `${request.viewer.self_description} ${request.viewer.connection_intent}`,
        },
      ]),
    ),
  };
}

File: server/evals/harness/fixtures.py
"""사례 로더. lib/fixtures.js 와 같은 결과를 내야 한다(후보 순서 포함)."""

from __future__ import annotations

import json
from pathlib import Path
from typing import Any

EVALS_DIR = Path(__file__).resolve().parent.parent
CASES_DIR = EVALS_DIR / "fixtures" / "cases"


def _read(relative: str) -> dict[str, Any]:
    return json.loads((EVALS_DIR / relative).read_text(encoding="utf-8"))


def _strip(pool: dict[str, Any]) -> dict[str, Any]:
    return {k: v for k, v in pool.items() if not k.startswith("_")}


TUNING_POOL = _strip(_read("fixtures/participants.json"))
HELDOUT_POOL = _strip(_read("fixtures/participants_heldout.json"))
# 프롬프트 동결 후 새로 만든 보류 풀. 기존 보류는 이미 결과를 본 뒤라 신선하지 않다.
FRESH_POOL = _strip(_read("fixtures/participants_fresh.json"))
_POOLS = {"tuning": TUNING_POOL, "heldout": HELDOUT_POOL, "fresh": FRESH_POOL}


def _fnv1a(text: str) -> int:
    """lib/fixtures.js 의 fnv1a 와 같은 값을 내야 한다(32비트)."""
    h = 0x811C9DC5
    for ch in text:
        h ^= ord(ch)
        h = (h * 0x01000193) & 0xFFFFFFFF
    return h


def _shuffle(candidates: list[dict], seed: str) -> list[dict]:
    # 사례 파일의 후보 순서에는 '기대 순위' 신호가 들어 있다. 그대로 두면 동점을 입력 순서로
    # 깨는 기준선이 판단 없이 1순위를 맞힌다. 고정 해시 순서로 그 신호를 지운다.
    #
    # 시드는 shuffle_group 이다. 교류 의도만 바꾼 비교 쌍은 후보 제시 순서까지 같아야
    # 순위 변화의 원인을 의도 변경으로 좁힐 수 있다.
    return sorted(
        candidates,
        key=lambda c: (_fnv1a(f"{seed}:{c['user_id']}"), c["user_id"]),
    )


def list_cases(case_set: str | None = None) -> list[str]:
    ids = sorted(p.stem for p in CASES_DIR.glob("*.json"))
    if case_set is None:
        return ids
    return [i for i in ids if _read(f"fixtures/cases/{i}.json")["set"] == case_set]


def load_case(case_id: str) -> dict[str, Any]:
    body = _read(f"fixtures/cases/{case_id}.json")
    pool = _POOLS.get(body["pool"], TUNING_POOL)

    def resolve(entry: dict, role: str) -> dict:
        base = pool.get(entry["ref"])
        if base is None:
            raise KeyError(f"{case_id}: {role} 참조 '{entry['ref']}'가 해당 풀에 없습니다.")
        merged = dict(base)
        merged.update({k: v for k, v in entry.items() if k != "ref"})
        return merged

    viewer = resolve(body["request"]["viewer"], "viewer")
    authored = [resolve(c, "candidate") for c in body["request"]["candidates"]]
    body["request"] = {
        "viewer": viewer,
        "candidates": _shuffle(authored, body.get("shuffle_group") or case_id),
        "policy": body["request"].get("policy") or {},
        "fault": body["request"].get("fault"),
        "case_id": case_id,
        "authored_candidate_order": [c["user_id"] for c in authored],
        "shuffle_group": body.get("shuffle_group") or case_id,
    }
    return body


def to_request(request: dict):
    """평가 사례를 모듈의 RecommendationRequest 로 바꾼다."""
    from app.ai import ParticipantProfile, RecommendationRequest

    def profile(raw: dict):
        return ParticipantProfile(
            user_id=raw["user_id"],
            self_description=raw.get("self_description", ""),
            connection_intent=raw.get("connection_intent", ""),
            profile_revision=raw.get("profile_revision", 1),
        )

    return RecommendationRequest(
        viewer=profile(request["viewer"]),
        candidates=tuple(profile(c) for c in request["candidates"]),
    )
